crates/ide-assists/src/handlers/convert_for_to_while_let.rs

```diff
@@ -0,0 +1,383 @@
+use hir::{sym, Name};
+use ide_db::famous_defs::FamousDefs;
+use syntax::{
+    ast::{
+        self,
+        edit::{AstNodeEdit, IndentLevel},
+        make, HasLoopBody,
+    },
+    AstNode,
+};
+
+use crate::{AssistContext, AssistId, AssistKind, Assists};
+
+// Assist: convert_for_loop_to_while_let
+//
+// Converts a for loop into a while let on the Iterator.
+//
+// ```
+// fn main() {
+//     let x = vec![1, 2, 3];
+//     for$0 v in x {
+//         let y = v * 2;
+//     };
+// }
+// ```
+// ->
+// ```
+// fn main() {
+//     let x = vec![1, 2, 3];
+//     while let Some(v) = x.into_iter().next() {
+//         let y = v * 2;
+//     };
+// }
+// ```
+pub(crate) fn convert_for_loop_to_while_let(
+    acc: &mut Assists,
+    ctx: &AssistContext<'_>,
+) -> Option<()> {
+    let for_loop = ctx.find_node_at_offset::<ast::ForExpr>()?;
+    let iterable = for_loop.iterable()?;
+    let pat = for_loop.pat()?;
+    let body = for_loop.loop_body()?;
+    if body.syntax().text_range().start() < ctx.offset() {
+        cov_mark::hit!(not_available_in_body);
+        return None;
+    }
+
+    acc.add(
+        AssistId("convert_for_loop_to_while_let", AssistKind::RefactorRewrite),
+        "Replace this for loop with `while let`",
+        for_loop.syntax().text_range(),
+        |builder| {
+            let iter = if let Some((expr_behind_ref, method)) =
+                is_ref_and_impls_iter_method(&ctx.sema, &iterable)
+            {
+                // We have either "for x in &col" and col implements a method called iter
+                //             or "for x in &mut col" and col implements a method called iter_mut
+                make::expr_method_call(
+                    expr_behind_ref,
+                    make::name_ref(method.as_str()),
+                    make::arg_list([]),
+                )
+            } else if let ast::Expr::RangeExpr(..) = iterable {
+                // range expressions need to be parenthesized for the syntax to be correct
+                make::expr_paren(iterable)
+            } else if impls_core_iter(&ctx.sema, &iterable) {
+                // We can't use `iterable` if its immutable, so we create a tmp mutable variable
+                let mut_expr = make::let_stmt(
+                    make::ident_pat(false, true, make::name("tmp")).into(),
+                    None,
+                    Some(iterable),
+                );
+                builder.insert(
```

Comment:
Less of a priority but would you be willing to use `SyntaxEditor`-based edits instead?
[`introduce_named_type_parameter.rs`](https://github.com/rust-lang/rust-analyzer/blob/8d01570b5e812a49daa1f08404269f6ea5dd73a1/crates/ide-assists/src/handlers/introduce_named_type_parameter.rs) has a small example of using `SyntaxEditor` along with a `SourceChangeBuilder`, but there are also other files you can look at for more examples.